hashicorp/dev-portal
# feat: tutorials opt in / opt out

## 🔗 Relevant links



- [Preview link](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/vault/tutorials) 🔎
- [Asana task](https://app.asana.com/0/1199634971449915/1201987217085805) 🎟️

## 🗒️ What

This PR adds the opt-out workflow for tutorials. A new `optInOutSlot` was added to the sidebar sidecar layout. That should render the opt out component, which shows a button and handles the dialog as well. 

This preview is linked to [this branch in Learn](https://github.com/hashicorp/learn/pull/4480), so you can click between the two. For release we'll need to update that base url. 

The goal of this work is to get the general opt-out flow in place and then handle any UI refinements later. 

## 📸 Design Screenshots

[See design here. ](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Dev-Portal?node-id=8356%3A92080) They also adjusted things to use a select dropdown for the dialog, [which can be seen here. ](https://www.figma.com/file/VD7ahvXuXWJApeGnhbW4hv/Dev-Portal?node-id=9090%3A123877)

## 🧪 Testing

- Visit [waypoint](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/waypoint/tutorials) and [vault](https://dev-portal-git-kstutorials-opt-in-hashicorp.vercel.app/vault/tutorials) tutorials pages. Verify that no opt out button renders since the opt-in hasn't been set yet. 
- Now visit this [learn preview to opt-in](https://learn-git-ksspike-opt-in-redirects-hashicorp.vercel.app/waypoint). That should redirect you back to this preview. 
- Click around to a tutorial or collection, hit opt out and open the dialog. Opt out to test the full flow. 

## 💭 Anything else?

The UI still needs to be polished a bit with the dialog but I want to get this core functionality in. I haven't styled the select options because I didn't have time and that feels like a _nice to have_. 

The opt out button can look weird on tutorial pages that have really long titles since we haven't implemented breadcrumb truncation yet. I'm not planning on addressing that in this PR. 

We will be [implementing the toast](https://app.asana.com/0/1201903760348480/1201979557016120) in a follow up PR. 

The form data doesn't submit to anywhere yet, that is follow up work.
